scalalib/src/JavaModule.scala

```diff
@@ -772,13 +772,32 @@ trait JavaModule
 
   def forkWorkingDir: Target[Path] = T { os.pwd }
 
+  /**
+   * @param hint "all" - fetch all dependencies,
+   *             "source" | "sources" - fetch source dependencies
+   */
   @nowarn("msg=pure expression does nothing")
-  override def prepareOffline(): Command[Unit] = T.command {
-    super.prepareOffline()()
-    resolvedIvyDeps()
-    zincWorker.prepareOffline()()
-    resolvedRunIvyDeps()
-    ()
+  override def prepareOffline(hint: String*): Command[Unit] = {
```

Comment:
This signature is kind of weird

- Do we need both `source` and `sources` as options? 
- Shouldn't passing in all three values at the same time e.g. `./mill foo.prepareOffline source sources all` throw an error?
- Should we use a flag `./mill foo.prepareOffline --all` `./mill foo.prepareOffline --sources`? 
- Or even just separate targets `./mill foo.prepareOffline` `./mill foo.prepareOfflineWithSources`?

Reply:
> This signature is kind of weird
> 
>* Do we need both `source` and `sources` as options?

No, either `source` or `sources` is enough.

>* Shouldn't passing in all three values at the same time e.g. `./mill foo.prepareOffline source sources all` throw an error?

I think not. 

> * Presumably invalid strings like `./mill foo.prepareOffline invalid-flag` should be an error too

Imagine a project with multiple different modules. If we reject all hints we don't support, we can never let other modules support new hints. E.g. To run `mill __.prepareOffline foo`  where hint is special to `BarModule` but only some modules are implemented with `extends BarModule` and all other modules should do default offline preparations.

>* Should we use a flag `./mill foo.prepareOffline --all` `./mill foo.prepareOffline --sources`?

That would be clearly the better choice, but I don't see how to implement it in Mill.

>* Or even just separate targets `./mill foo.prepareOffline` `./mill foo.prepareOfflineWithSources`?

Yeah, this would be the natural way to make it right, but then we don't need the `OfflineSupportModule` at all.  It would result in many `prepareOffline*` targets and we don't have any support to call them efficiently. In the end, the user needs to learn many different targets, but not a single one.

Reply:
I think I see what you're trying to do with the hints. It's still rather odd though; they're kind of separate targets but not really, so there's no way to inspect them or discover them. `./mill resolve foo._` won't show you which hints are valid, and as you said invalid hints won't throw an exception. Overall doesn't seem like a good setup to me, apart from not really fitting into how any of the other mill modules/targets/commands work.

Here's one idea: what if we made each module have an `object prepareOffline extends Module{}`, in which we would put whatever `def all = T{...}` `def sources = T{...}` targets we would like? Then people would select what they want via `./mill foo.prepareOffline.all`, or even across multiple targets with multiple kinds of hint via `./mill __.prepareOffline.{all,sources}`, while still letting people do things like `./mill resolve foo.prepareOffline._` to see what's available.

If we're worried about the `object prepareOffline` making it hard to override the tasks, we could do something like 

```scala
val prepareOffline: PrepareOfflineModule = new PrepareOfflineModule{}
trait PrepareOfflineModule extends Module{...}
``` 

to allow subclasses to subclass `trait PrepareOfflineModule` and override `val prepareOffline` if necessary

Reply:
Nice trick, but I honestly think, this is overkill. I tend to just introduce a binary `all` flag to the target. If we need more, we have to introduce new targets though, but that true for your suggestion as well.